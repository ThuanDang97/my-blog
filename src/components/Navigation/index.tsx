import { Box, List, ListItem, SvgIcon, useTheme } from '@mui/material'
import Link from 'next/link'
import { lazy, memo, Suspense, useState } from 'react'
import SearchIcon from '@mui/icons-material/Search'

// Components
import Button from '@components/Button'

const Search = lazy(() => import('@components/Search'))

type TNavItem = {
  id: number
  name: string
  path: string
}

export interface INavigation {
  navItems: TNavItem[]
}

const Navigation = ({ navItems }: INavigation) => {
  const theme = useTheme()
  const [isOpen, setIsOpen] = useState(false)

  const handleOpen = () => {
    setIsOpen(true)
  }

  const handleClose = () => {
    setIsOpen(false)
  }

  const handleGetValue = () => {
    console.log(123)
  }

  const handleKeyDown = () => {
    console.log(123)
  }

  return (
    <>
      <Box
        display="flex"
        justifyContent="space-between"
        padding="12px 12px 12px 25px"
        width="100%"
        alignItems="center"
        boxShadow={`${theme.palette.grey[400]} 0px 3px 8px`}
        bgcolor={theme.palette.common.white}
      >
        <List
          sx={{
            display: 'flex',
            gap: '30px',
          }}
        >
          {navItems.map((item) => (
            <ListItem
              key={item.id}
              sx={{
                padding: 0,
                fontSize: '18px',
                textTransform: 'uppercase',
                fontWeight: 600,
                letterSpacing: '2px',
                width: 'auto',
                ':hover': {
                  color: theme.palette.grey[500],
                },
              }}
            >
              <Link href={item.path} prefetch={false}>
                {item.name}
              </Link>
            </ListItem>
          ))}
        </List>

        {/* Search Button */}
        <Button
          onClick={handleOpen}
          variant="text"
          sx={{
            backgroundColor: 'transparent',
            ':hover': {
              backgroundColor: 'transparent',
            },
          }}
        >
          <SvgIcon
            sx={{
              color: theme.palette.common.black,
              width: 24,
              height: 24,
              cursor: 'pointer',
              ':hover': {
                color: theme.palette.grey[500],
              },
            }}
          >
            <SearchIcon />
          </SvgIcon>
        </Button>
      </Box>

      {/* Modal Search */}
      <Suspense fallback={<div>loading.....</div>}>
        <Search
          isOpen={isOpen}
          onClose={handleClose}
          onChange={handleGetValue}
          onKeyDown={handleKeyDown}
        />
      </Suspense>
    </>
  )
}

export default memo(Navigation)
