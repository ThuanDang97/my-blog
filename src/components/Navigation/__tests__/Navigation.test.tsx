// Utils
import { fireEvent, render, screen } from '@utils/testUtils'

// Mocks data
import { MOCK_NAVIGATE } from '@constants/__mock__/mockData'

// Components
import Navigation, { INavigation } from '..'

describe('Navigation Component', () => {
  const mockProps: INavigation = MOCK_NAVIGATE

  jest.mock('next/router', () => ({
    useRouter: () => ({
      push: jest.fn(),
      back: jest.fn(),
      events: {
        on: jest.fn(),
        off: jest.fn(),
        emit: jest.fn(),
      },
    }),
  }))

  it('Should show match Navigation DOM Snapshot ', () => {
    const { container } = render(<Navigation {...mockProps} />)

    expect(container).toMatchSnapshot()
  })

  it('should render math content Navigation', () => {
    render(<Navigation {...mockProps} />)

    const home = screen.getByRole('link', {
      name: /home/i,
    })
    const mostView = screen.getByRole('link', {
      name: /Most View/i,
    })
    const blog = screen.getByRole('link', {
      name: /Blog/i,
    })
    const about = screen.getByRole('link', {
      name: /About/i,
    })

    expect(home).toHaveTextContent('Home')
    expect(mostView).toHaveTextContent('Most View')
    expect(blog).toHaveTextContent('Blog')
    expect(about).toHaveTextContent('About')

    expect(home).toHaveAttribute('href', '/')
    expect(mostView).toHaveAttribute('href', '/most-view')
    expect(blog).toHaveAttribute('href', '/blog')
    expect(about).toHaveAttribute('href', '/about')
  })

  it('handleOpen should call and open modal search', () => {
    render(<Navigation {...mockProps} />)

    const buttonSearch = screen.getByRole('button', {
      name: /search/i,
    })

    fireEvent.click(buttonSearch)

    expect(screen.getByTestId('backdrop')).toBeInTheDocument()
  })

  it('handleClose should call and close modal search', () => {
    const { container } = render(<Navigation {...mockProps} />)

    const buttonSearch = screen.getByRole('button', {
      name: /search/i,
    })

    fireEvent.click(buttonSearch)

    const buttonClose = screen.getByTestId('close-icon')

    fireEvent.click(buttonClose)

    expect(container).not.toBe('Type to search')
  })

  it('handleGetValue should call and get correct value', () => {
    render(<Navigation {...mockProps} />)

    const buttonSearch = screen.getByRole('button', {
      name: /search/i,
    })

    fireEvent.click(buttonSearch)

    const search = screen.getByRole('textbox') as HTMLInputElement

    fireEvent.change(search, {
      target: { value: 'Search Value' },
    })

    expect(search.value).toBe('Search Value')
  })

  it('handleKeyDown should call and get correct value', () => {
    render(<Navigation {...mockProps} />)

    const buttonSearch = screen.getByRole('button', {
      name: /search/i,
    })

    fireEvent.click(buttonSearch)

    const search = screen.getByRole('textbox')
    fireEvent.keyDown(search)

    expect(search).toBeTruthy()
  })
})
