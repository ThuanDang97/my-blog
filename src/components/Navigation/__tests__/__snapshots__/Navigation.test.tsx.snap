// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation Component Should show match Navigation DOM Snapshot  1`] = `
<div>
  <div
    class="MuiBox-root css-7w1jqr"
  >
    <ul
      class="MuiList-root MuiList-padding css-1mq3yj5-MuiList-root"
    >
      <li
        class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1rk1jkp-MuiListItem-root"
      >
        <a
          href="/"
        >
          Home
        </a>
      </li>
      <li
        class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1rk1jkp-MuiListItem-root"
      >
        <a
          href="/most-view"
        >
          Most View
        </a>
      </li>
      <li
        class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1rk1jkp-MuiListItem-root"
      >
        <a
          href="/blog"
        >
          Blog
        </a>
      </li>
      <li
        class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1rk1jkp-MuiListItem-root"
      >
        <a
          href="/about"
        >
          About
        </a>
      </li>
    </ul>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-15on3wz-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1unfims-MuiSvgIcon-root"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="SearchIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
          />
        </svg>
      </svg>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
  <div>
    loading.....
  </div>
</div>
`;
